zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/serializer/StringSerializer.java

```diff
@@ -0,0 +1,80 @@
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.gateway.rest.serializer;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import io.camunda.zeebe.gateway.rest.RequestMapper;
+import java.io.IOException;
+import org.springframework.http.HttpHeaders;
+import org.springframework.web.context.request.RequestAttributes;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+public class StringSerializer extends JsonSerializer<String> {
+
+  public static final String PACKAGE_GATEWAY_REST = "io.camunda.zeebe.gateway.protocol.rest";
+  public static final String KEY = "Key";
+  public static final String KEYS = "Keys";
+
+  private void writeDefault(final String value, final JsonGenerator gen) throws IOException {
+    writeString(value, gen);
+  }
+
+  private void writeString(final String value, final JsonGenerator gen) throws IOException {
+    gen.writeString(value);
+  }
+
+  private void writeNumber(final String value, final JsonGenerator gen) throws IOException {
+    gen.writeNumber(Long.parseLong(value));
+  }
+
+  private boolean gatewayRestPackage(final Object value) {
+    return value != null && value.getClass().getPackageName().startsWith(PACKAGE_GATEWAY_REST);
+  }
+
+  private boolean gatewayRestPackage(final JsonGenerator gen) {
+    return gatewayRestPackage(gen.getOutputContext().getCurrentValue())
+        || gatewayRestPackage(gen.getOutputContext().getParent().getCurrentValue());
+  }
+
+  private boolean fieldNameEndsWithKey(final JsonGenerator gen) {
+    final var currentName = gen.getOutputContext().getCurrentName();
+    final var parentName = gen.getOutputContext().getParent().getCurrentName();
+    return currentName != null && (currentName.equals("key") || currentName.endsWith(KEY))
+        || parentName != null && parentName.endsWith(KEYS);
+  }
+
+  @Override
+  public void serialize(
+      final String value, final JsonGenerator gen, final SerializerProvider serializers)
+      throws IOException {
+
+    if (!gatewayRestPackage(gen) || !fieldNameEndsWithKey(gen)) {
+      writeString(value, gen);
+      return;
+    }
+
+    final RequestAttributes requestAttributes = RequestContextHolder.getRequestAttributes();
+    if (requestAttributes instanceof final ServletRequestAttributes r) {
+      final var acceptHeader = r.getRequest().getHeader(HttpHeaders.ACCEPT);
+
+      if (acceptHeader != null && acceptHeader.equals(RequestMapper.MEDIA_TYPE_KEYS_NUMBER_VALUE)) {
+        writeNumber(value, gen);
+        return;
+      }
```

Comment:
💭 One issue we'll run into with this type of handling values is that we have fields that are ending in `key` and are always `string` values. They must not be turned into numbers. One example is the `correlationKey` in message endpoints. This is why it works with the LongSerializer in 8.6: we will write number values as is if needed, but we can also turn them into strings if required. It doesn't work the other way around.

Reply:
Ah okay very good point, that's not so good. I'll think about this then, but looks like we won't have such a clean solution as with longs in 8.6...